### PR TITLE
Contracts & Harnesses for `non_null::sub` and `non_null::sub_ptr` and `non_null::offset_from`

### PR DESCRIPTION
**Pull request by Jimmycreative**
Towards https://github.com/model-checking/verify-rust-std/issues/53

Changes
added contract and harness for non_null::sub
added contract and harness for non_null::sub_ptr

Revalidation
To revalidate the verification results, run kani verify-std -Z unstable-options "path/to/library"  -Z function-contracts -Z mem-predicates --harness ptr::non_null::verify This will run both harnesses. All default checks should pass:

```
SUMMARY:
 ** 0 of 1622 failed

VERIFICATION:- SUCCESSFUL
Verification Time: 0.3814842s

SUMMARY:
 ** 0 of 1780 failed (1 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 0.44192737s

Complete - 2 successfully verified harnesses, 0 failures, 2 total.
``` 

### Clarifying Questions 
The proof now only handles the array with a fixed size and uses a random element in the arr for subtraction. The element is i32 type. Is this ok for the current stage? Or maybe we need to consider other types such as i64, etc and maybe change the arr to a bigger size?

### CODE REVIEW

File: library/core/src/ptr/non_null.rs [hunk 1]
```diff
@@ -622,6 +627,11 @@ impl<T: ?Sized> NonNull<T> {
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_allow_const_fn_unstable(unchecked_neg)]
+    #[kani::requires(!self.as_ptr().is_null())]
```
Comment (feliperodri) [inline library/core/src/ptr/non_null.rs:630]: All CI failures are related of using contracts directly from the Kani crate.
```
^^^^ use of undeclared crate or module `kani`
```
The annotations should be tool agnostic, so we have a wrapper. Please replace all contracts to use the wrapper `safety::requires`. Take a look at this [pull request](https://github.com/model-checking/verify-rust-std/pull/69/files) as an example.

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: library/core/src/ptr/non_null.rs (hunk 1 above)
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs]: Removed thanks!

File: library/core/src/ptr/non_null.rs [hunk 2]
```diff
@@ -1770,3 +1781,56 @@ impl<T: ?Sized> From<&T> for NonNull<T> {
...
+        let arr: [i32; SIZE] = kani::any();  
+        // Get a raw pointer to the array
+        let raw_ptr: *mut i32 = arr.as_ptr() as *mut i32;  
+        // NonNUll pointer to the random offset
+        let ptr = unsafe { NonNull::new(raw_ptr.add(offset)).unwrap() };  
+        // Create a non-deterministic count value
+        let count: usize = kani::any();  
+ 
+        // SAFETY: Ensure that the subtraction does not go out of the bounds of the array
+        kani::assume(count < SIZE - offset);
```
Comment (feliperodri) [inline library/core/src/ptr/non_null.rs:1805]: Why do you need this assumption in the harness?

File: library/core/src/ptr/non_null.rs [hunk 3]
```diff
@@ -1770,3 +1781,56 @@ impl<T: ?Sized> From<&T> for NonNull<T> {
...
+
+        unsafe {
+            // Perform the pointer subtraction from the last element
+            let result = ptr.sub(count);
+        }
+    }
+    
+    #[kani::proof_for_contract(NonNull::sub_ptr)]
+    pub fn non_null_check_sub_ptr() {
```
Comment (feliperodri) [inline library/core/src/ptr/non_null.rs:1814]: Same comments as above.

**Review by feliperodri — DISMISSED** [review]

File: library/core/src/ptr/non_null.rs [hunk 4]
```diff
@@ -1770,3 +1781,56 @@ impl<T: ?Sized> From<&T> for NonNull<T> {
...
 }
+
+#[cfg(kani)]
+#[unstable(feature="kani", issue="none")]
+mod verify {
+    use super::*;
+
+    #[kani::proof_for_contract(NonNull::sub)]
+    pub fn non_null_check_sub() {
+        const SIZE: usize = 10;
```
[the remark this answers is not in the document]
Comment (Jimmycreative) [reply library/core/src/ptr/non_null.rs:1792]: Currently just use 10 for fixed size as an example. Do you suggest to use a larger size?

File: library/core/src/ptr/non_null.rs [hunk 5]
```diff
@@ -622,6 +626,11 @@ impl<T: ?Sized> NonNull<T> {
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_allow_const_fn_unstable(unchecked_neg)]
+    #[requires(!self.as_ptr().is_null())]
+    #[requires(count <= isize::MAX as usize)] // SAFETY: count does not overflow isize
```
[the remark this answers is not in the document]
Comment (Jimmycreative) [reply library/core/src/ptr/non_null.rs:630]: Hi, just to clarify: if the safety comment states this, does it mean that self is always non-null and count is always smaller than isize::MAX when calling the function?

File: library/core/src/ptr/non_null.rs [hunk 6]
```diff
@@ -846,6 +855,8 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[unstable(feature = "ptr_sub_ptr", issue = "95892")]
     #[rustc_const_unstable(feature = "const_ptr_sub_ptr", issue = "95892")]
+    #[requires(!self.as_ptr().is_null())]
+    #[requires(!subtracted.as_ptr().is_null())]
```
[the remark this answers is not in the document]
Comment (Jimmycreative) [reply library/core/src/ptr/non_null.rs:859]: I added the line below:

```rust
#[kani::ensures(|result: &usize| *result == self.as_ptr().offset_from(subtracted.as_ptr()) as usize)]
```

File: library/core/src/ptr/non_null.rs (hunk 2 above)
Comment (Jimmycreative) [reply library/core/src/ptr/non_null.rs:1805]: without this line it seems it will be out of boundary since I got the below errors:

```rust
 ** 2 of 1741 failed (1 unreachable)
Failed Checks: offset: attempt to compute number in bytes which would overflow
Failed Checks: attempt to multiply with overflow
```

File: library/core/src/ptr/non_null.rs (hunk 3 above)
Comment (Jimmycreative) [reply library/core/src/ptr/non_null.rs:1814]: Sorry, which line has the problem?

File: .gitignore
Comment (danielhumanmod) [inline .gitignore:1]: Remove the .gitignore file

File: library/core/src/ptr/non_null.rs [hunk 7]
```diff
@@ -10,7 +10,6 @@ use crate::ub_checks::assert_unsafe_precondition;
 use crate::{fmt, hash, intrinsics, ptr};
 use safety::{ensures, requires};
 
-
```
Comment (danielhumanmod) [inline library/core/src/ptr/non_null.rs:13]: NIT: unnecessary white space

File: library/core/src/ptr/non_null.rs [hunk 8]
```diff
@@ -631,6 +630,9 @@ impl<T: ?Sized> NonNull<T> {
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_allow_const_fn_unstable(unchecked_neg)]
+    #[kani::requires(count.checked_mul(core::mem::size_of::<T>()).is_some())] // Prevent offset overflow
```
Comment (danielhumanmod) [inline library/core/src/ptr/non_null.rs:633]: check the ptr is non null as well?

File: .gitignore [hunk 9]
```diff
@@ -49,4 +49,4 @@ package-lock.json
 #
 # already existing elements were commented out
 
-#/target
+#/target
```
Comment (carolynzech) [inline .gitignore:52]: Please undo this change (put the newline back).

File: library/core/src/ptr/non_null.rs [hunk 10]
```diff
@@ -1,6 +1,6 @@
 use crate::cmp::Ordering;
 use crate::marker::Unsize;
-use crate::mem::{MaybeUninit, SizedTypeProperties};
+use crate::mem::{self, MaybeUninit, SizedTypeProperties};
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:3]: Why this import?

File: library/core/src/ptr/non_null.rs (hunk 2 above)
Comment (carolynzech) [reply library/core/src/ptr/non_null.rs:1826]: Instead of encoding this as an assumption in the harness, you should write them as preconditions (`#[requires(...)]`).
You can think of a contract as saying: "Given that the preconditions are met, the postconditions hold." So, if your postconditions (your `ensures` contracts) hold if and only if some preconditions (like this one) do, you should write those preconditions using `requires`.
A good rule of thumb is to try to minimize the `kani::assume` calls present in your harnesses--if you're using them to encode a precondition, they should go in the contracts instead. One exception to this rule of thumb would be if we decide to use `kani::assume` to restrict the size of the input space for harness performance reasons.

**Review by carolynzech — CHANGES_REQUESTED** [review]

File: library/core/src/ptr/non_null.rs [hunk 11]
```diff
@@ -631,6 +630,9 @@ impl<T: ?Sized> NonNull<T> {
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_allow_const_fn_unstable(unchecked_neg)]
+    #[kani::requires(count.checked_mul(core::mem::size_of::<T>()).is_some())] // Prevent offset overflow
+    #[kani::requires(count * core::mem::size_of::<T>() <= isize::MAX as usize)]
+    #[kani::ensures(|result: &NonNull<T>| result.as_ptr() == self.as_ptr().offset(-(count as isize)))]
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:635]: add post-condition that asserts the resulted pointer is pointing to the same object as the original pointer

File: library/core/src/ptr/non_null.rs [hunk 12]
```diff
@@ -762,6 +764,7 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[kani::ensures(|result: &isize| *result == (self.as_ptr() as isize - origin.as_ptr() as isize) / core::mem::size_of::<T>() as isize)]
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:767]: add post-condition that asserts the resulted pointer is pointing to the same object as the original pointer

File: library/core/src/ptr/non_null.rs [hunk 13]
```diff
@@ -855,6 +858,7 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[unstable(feature = "ptr_sub_ptr", issue = "95892")]
     #[rustc_const_unstable(feature = "const_ptr_sub_ptr", issue = "95892")]
+    #[kani::ensures(|result: &usize| *result == self.as_ptr().offset_from(subtracted.as_ptr()) as usize)]
```
Comment (QinyuanWu) [inline library/core/src/ptr/non_null.rs:861]: add post-condition that asserts the resulted pointer is pointing to the same object as the original pointer

**Review by QinyuanWu — CHANGES_REQUESTED** [review]

**Review by Dhvani-Kapadia — COMMENTED** [review]
For sub_ptr, it is mentioned in the description that it has the added precondition that offset is guaranteed to be non-negative, should that condition be in the requires clause of the contract?

File: library/core/src/ptr/non_null.rs (hunk 2 above)
Comment (Jimmycreative) [reply library/core/src/ptr/non_null.rs]: this has been resolved using the same allocation api.

File: library/core/src/ptr/non_null.rs [hunk 14]
```diff
@@ -631,6 +631,12 @@ impl<T: ?Sized> NonNull<T> {
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_allow_const_fn_unstable(unchecked_neg)]
+    #[requires(
+        count.checked_mul(core::mem::size_of::<T>()).is_some() &&
+        count * core::mem::size_of::<T>() <= isize::MAX as usize && 
+        kani::mem::same_allocation(self.as_ptr() as *const(), self.as_ptr().sub(count) as *const())
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:637]:
```suggestion
        kani::mem::same_allocation(self.as_ptr(), self.as_ptr().wrapping_sub(count))
```

Comment (carolynzech) [reply library/core/src/ptr/non_null.rs:637]: Suggesting `wrapping_sub` because `sub` will invoke UB if we cross object boundaries, so if you execute `sub` and it's not in the same allocation, the contract will cause UB. Instead, use `wrapping_sub` so that it produces a pointer, and then `same_allocation` can reason about whether that pointer indeed refers to the same allocation as `self`.

File: library/core/src/ptr/non_null.rs [hunk 15]
```diff
@@ -762,6 +768,7 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[ensures(|result: &isize| *result == (self.as_ptr() as isize - origin.as_ptr() as isize) / core::mem::size_of::<T>() as isize)]
     pub const unsafe fn offset_from(self, origin: NonNull<T>) -> isize
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:772]: Can you add some safety preconditions for this function? https://doc.rust-lang.org/beta/std/ptr/struct.NonNull.html#method.offset_from

Comment (carolynzech) [reply library/core/src/ptr/non_null.rs:772]: Also, these docs say that the function panics if `T` is a ZST, so can you add a [`#[should_panic]` harness ](https://model-checking.github.io/kani/reference/attributes.html?highlight=panic#kanishould_panic)checking that?

File: library/core/src/ptr/non_null.rs [hunk 16]
```diff
@@ -855,6 +862,7 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[unstable(feature = "ptr_sub_ptr", issue = "95892")]
     #[rustc_const_unstable(feature = "const_ptr_sub_ptr", issue = "95892")]
+    #[ensures(|result: &usize| *result == self.as_ptr().offset_from(subtracted.as_ptr()) as usize)]
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:865]: Can you add safety preconditions? https://doc.rust-lang.org/beta/std/ptr/struct.NonNull.html#method.sub_ptr

File: library/core/src/ptr/non_null.rs [hunk 17]
```diff
@@ -1803,4 +1811,67 @@ mod verify {
         let maybe_null_ptr =  if kani::any() { xptr as *mut i32 } else { null_mut() };
         let _ = NonNull::new(maybe_null_ptr);
     }
+
+    #[kani::proof_for_contract(NonNull::sub)]
+    pub fn non_null_check_sub() {
+        const SIZE: usize = 10;
+        // Randomiz pointer offset within array bound
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:1818]: Can you use the `PointerGenerator` functionality for your harnesses instead? I think it will be cleaner -- see [here](https://github.com/model-checking/verify-rust-std/discussions/162).

**Review by carolynzech — COMMENTED** [review]
Can you go through the other comments you've received above and mark them as resolved (if you've indeed resolved them)? We want to have everything resolved before approving.

File: library/core/src/ptr/non_null.rs (hunk 14 above)
Comment (Jimmycreative) [reply library/core/src/ptr/non_null.rs]: resolved

File: library/core/src/ptr/non_null.rs (hunk 15 above)
Comment (Jimmycreative) [reply library/core/src/ptr/non_null.rs:779]: all done

File: library/core/src/ptr/non_null.rs [hunk 18]
```diff
@@ -762,6 +768,14 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[kani::should_panic]
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:771]: This attribute should be on the harness and not the function itself.

File: library/core/src/ptr/non_null.rs [hunk 19]
```diff
@@ -762,6 +768,14 @@ impl<T: ?Sized> NonNull<T> {
...
+    #[requires(
+        (self.as_ptr() as usize == origin.as_ptr() as usize) ||
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:773]: This case is included in `same_allocation` so should not be needed.

File: library/core/src/ptr/non_null.rs [hunk 20]
```diff
@@ -762,6 +768,14 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[kani::should_panic]
+    #[requires(
+        (self.as_ptr() as usize == origin.as_ptr() as usize) || 
+        (kani::mem::same_allocation(self.as_ptr(), origin.as_ptr()) && 
+         ((self.as_ptr() as usize >= origin.as_ptr() as usize) &&
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:775]: I don't think the function has this requirement. The result can be positive (if `origin` is less than or equal to `self`) and negative otherwise.

File: library/core/src/ptr/non_null.rs [hunk 21]
```diff
@@ -1803,4 +1820,56 @@ mod verify {
...
+        let other_ptr = unsafe { NonNull::new(raw_ptr2).unwrap() };
+
+        // Perform pointer subtraction safely
+        unsafe {
+            let distance = ptr.sub_ptr(other_ptr);
+        }
+    }
+
+    #[kani::proof_for_contract(NonNull::offset_from)]
+    pub fn non_null_check_offset_from() {
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1859]: We should use two `PointerGenerator` objects in this harness, similar to https://github.com/model-checking/verify-rust-std/pull/103/files#diff-430db13219f4c1ac4f4d35f685a45778572e039dab837c5ffeec9a810fc547b4R1859

Comment (Jimmycreative) [reply library/core/src/ptr/non_null.rs:1866]: I also modified the sub_ptr function to utilize two PointerGenerator instances.

File: library/core/src/ptr/non_null.rs [hunk 22]
```diff
@@ -762,6 +768,11 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires( 
+        (kani::mem::same_allocation(self.as_ptr(), origin.as_ptr()) && 
+        ((self.as_ptr() as usize - origin.as_ptr() as usize) % core::mem::size_of::<T>() == 0))
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:773]:
```suggestion
        ((self.as_ptr().addr() - origin.as_ptr().addr()) % core::mem::size_of::<T>() == 0))
```

File: library/core/src/ptr/non_null.rs [hunk 23]
```diff
@@ -855,6 +866,9 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[unstable(feature = "ptr_sub_ptr", issue = "95892")]
     #[rustc_const_unstable(feature = "const_ptr_sub_ptr", issue = "95892")]
+    #[requires(kani::mem::same_allocation(self.as_ptr(), subtracted.as_ptr()))] // Ensure both pointers are in the same allocation
+    #[requires((self.as_ptr() as usize) >= (subtracted.as_ptr() as usize))] // Ensure distance between pointers is non-negative
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:870]:
```suggestion
    #[requires((self.as_ptr().addr()) >= (subtracted.as_ptr().addr()))] // Ensure distance between pointers is non-negative
     #[requires((self.as_ptr().addr() - origin.as_ptr().addr()) % core::mem::size_of::<T>() == 0)]
```

`sub_ptr` has the same preconditions as `offset_from`, so added the exact multiples precondition.

File: library/core/src/ptr/non_null.rs [hunk 24]
```diff
@@ -1803,4 +1817,75 @@ mod verify {
...
+        let mut generator = kani::PointerGenerator::<SIZE>::new();
+        // Get a raw pointer from the generator within bounds
+        let raw_ptr: *mut i32 = generator.any_in_bounds().ptr;
+        // Create a non-null pointer from the raw pointer
+        let ptr = unsafe { NonNull::new(raw_ptr).unwrap() };
+        // Create a non-deterministic count value
+        let count: usize = kani::any();
+
+        unsafe {
+            // Perform the pointer subtraction from the last element
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:1833]:
```suggestion
```
Not sure what you mean by "the last element"-- `ptr` can refer to any position in the generator's buffer.

**Review by carolynzech — APPROVED** [review]
LGTM modulo these changes. Thanks!

File: library/core/src/ptr/non_null.rs [hunk 25]
```diff
@@ -887,6 +898,10 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[unstable(feature = "ptr_sub_ptr", issue = "95892")]
     #[rustc_const_unstable(feature = "const_ptr_sub_ptr", issue = "95892")]
+    #[requires(kani::mem::same_allocation(self.as_ptr(), subtracted.as_ptr()))] // Ensure both pointers are in the same allocation
+    #[requires((self.as_ptr().addr()) >= (subtracted.as_ptr().addr()))] // Ensure distance between pointers is non-negative
+     #[requires((self.as_ptr().addr() - origin.as_ptr().addr()) % core::mem::size_of::<T>() == 0)]
```
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:903]:
```suggestion
    #[requires((self.as_ptr().addr() - substracted.as_ptr().addr()) % core::mem::size_of::<T>() == 0)]
```

**Review by zhassan-aws — APPROVED** [review]

**Review by carolynzech — COMMENTED** [review]
trigger merge workflow

**Review by zhassan-aws — APPROVED** [review]
